_posts/2016-06-27-bits-of-shapeless-2.md

```diff
@@ -0,0 +1,519 @@
+
+```scala
+Witness(1)
+// res9: shapeless.Witness.Aux[Int(1)] = shapeless.Witness$$anon$1@19faedb1
+
+Witness("hello")
+// res10: shapeless.Witness.Aux[String("hello")] = shapeless.Witness$$anon$1@5fc01e92
+```
+
+Notice the types of these expressions, they refine integers and other literal types.
+
+And they can be put to work:
+```scala
+// A function that only accepts the integer 2:
+def just2(arg: Witness.`2`.T) = arg
+// just2: (arg: Int(2))Int
+
+just2(2)
+// res12: Int = 2
+```
+
+It fails, at compile-time, for every integer except 2.
+
+```scala
+just2(3)
+// <console>:23: error: type mismatch;
+//  found   : Int(3)
+//  required: Int(2)
+//        just2(3)
+//              ^
+
+just2(0)
+// <console>:23: error: type mismatch;
+//  found   : Int(0)
+//  required: Int(2)
+//        just2(0)
+//              ^
+```
+
+
+This `Witness.<literal>` weird syntax is just Scala's [dynamic member lookup][dynamic] along with more macro-based machinery. Details [here][singletons].
+
+Last but not the least look at the signature of the following function:
+
+```scala
+def singletonValue[K <: Symbol](implicit ev: Witness.Aux[K]): ev.T = ev.value
+// singletonValue: [K <: Symbol](implicit ev: shapeless.Witness.Aux[K])ev.T
+
+singletonValue[Witness.`'name`.T]
+// res15: Symbol with shapeless.tag.Tagged[String("name")] = 'name
+```
+
+
+This expression computes the `Witness` and singleton type for the symbol `'name`. If we associate a value and its type with a singleton typed symbol, we have a good starting point to handle case classes in a Generic way.
+
+
+### Labelling values with types
+
+As you may already have guessed, shapeless provides exactly what we need to label our values:
+
+```scala
+Object Labelled {
+//...
+  type FieldType[K, +V] = V with KeyTag[K, V]
+  trait KeyTag[K, +V]
+//...
+}
+```
+
+`FieldType[K,V]` specialises the type `V`, tagging it with the `K` type. And `K`, can be a  symbol, tagged with the literal symbol's name. Actually `K`is a singleton type which covers what we need to make a better Generic representation of case classes (and other types):
+
+```scala
+case class Person(name: String, age: Int)
+// defined class Person
+
+val p1 = Person("John", 30)
+// p1: Person = Person(John,30)
+
+val genPerson = LabelledGeneric[Person]
+// genPerson: shapeless.LabelledGeneric[Person]{type Repr = shapeless.::[String with shapeless.labelled.KeyTag[Symbol with shapeless.tag.Tagged[String("name")],String],shapeless.::[Int with shapeless.labelled.KeyTag[Symbol with shapeless.tag.Tagged[String("age")],Int],shapeless.HNil]]} = shapeless.LabelledGeneric$$anon$1@22115370
+
+val reprP1 = genPerson.to(p1)
+// reprP1: genPerson.Repr = John :: 30 :: HNil
+```
+
+Pay attention to the `Repr`type of the last expression. It is essentially a`HList` of the case class elements, each one refined wit the `FieldType` tagging type described above. Shapeless calls this a `Record`, and it has many interesting operations: by itself
```

Comment:
you are not referring the last expression, but the penultimate expression. at least the better way to check Repr type is on the penultimate one